gnosis/cowswap
# 130/cancellation pending

# Summary

Cancellation is pending until it's really cancelled.
Or at lest we hope it's after 1min :)

- When cancelling an order, change the activity status to `cancelling`
    - New pill color for `cancelling` orders
    - `cancel` button not visible for `cancelling` orders
    - No toast notification at this point
- Order status remains `pending` with a flag indicating it's being cancelled
- Re-worked `EventUpdater` to aggregate all state transitions, not only `pending` -> `fulfilled`
    - All pending orders are checked against API, and classified into 5 states:
        1. `unknown` - order not found on api
        2. `fulfilled` - there was a match
        3. `cancelled` - cancelled flag is set and it has been over 60s since order creation
        4. `expired` - it has been over 60s since order has expired
        5. `pending` - none of the above
    - For every order in one of the final states (`fulfilled`, `cancelled`, `expired`) a toast notification is shown
- Removed `ExpiredOrderWatcher` updater

## Activity status
![screenshot_2021-06-08_15-03-11](https://user-images.githubusercontent.com/43217/121401780-0117e380-c90e-11eb-9479-0d1cce6837e2.png)
1. Pending order. Has a cancel button
2. Cancelling order. Slightly different pill color. No cancel button
3. Cancelled order. Grey pill color.

# Testing

1. Place an order
2. Open the account details modal (click on `Pending`)
3. Click on `cancel` (on the right side of the pending order)
- [ ] On the modal, verify the order matches the one you want to cancel
4. Click on `Request cancellation`
- [ ] On your wallet, confirm you the order id matches the one you want to cancel
5. Sign the message
- [ ] Back at the account details modal, verify the order you requested cancellation is still pending, but no longer has a `cancel` button
- [ ] After a few seconds (up to 1min), verify a order cancellation pop up is displayed matching the cancelled order
- [ ] Open the GP Explorer link and verify the status is `Canceled`
